### PR TITLE
[pigeon] fixed error returning errors from async void(void) functions

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/84551

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/pigeon/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.2.3
```
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:1]: Optional nit: Dart pre-release semver shifts everything to the right, so a bugfix would generally be 0.2.2+1; this is a feature version bump.

File: packages/pigeon/lib/generator_tools.dart [hunk 2]
```diff
@@ -8,7 +8,7 @@ import 'dart:mirrors';
 import 'ast.dart';
 
 /// The current version of pigeon. This must match the version in pubspec.yaml.
```
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:10]: Do the pigeon automated tests check this in CI? If not, that would be a good thing to implement.

File: packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AsyncHandlersTest.m [hunk 3]
```diff
@@ -76,6 +77,10 @@ - (void)calculate:(Value* _Nullable)input
   }
 }
 
+- (void)voidVoid:(nonnull void (^)(FlutterError* _Nullable))completion {
```
Comment (stuartmorgan) [inline packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AsyncHandlersTest.m:80]: This name is... unusual. Is this following some defined convention?

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AsyncHandlersTest.m (hunk 3 above)
Comment (gaaclarke) [reply packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AsyncHandlersTest.m:80]: Yea, if you check out the pigeon file it means a function that takes no parameters and has no return type `void foo(void)`.

File: packages/pigeon/lib/generator_tools.dart (hunk 2 above)
Comment (gaaclarke) [reply packages/pigeon/lib/generator_tools.dart:10]: Yep, we have a test for that https://github.com/gaaclarke/packages/blob/pigeon-with-analyzer/packages/pigeon/test/version_test.dart#L11:L11

File: packages/pigeon/CHANGELOG.md (hunk 1 above)
Comment (gaaclarke) [reply packages/pigeon/CHANGELOG.md:1]: Yea, I know.  I've just been treating each fix as a minor release.  The nuance didn't seem that important leading up to 1.0.

**Pull request merged**